POSTGRES_XLOG_DIR to specify transaction log dir

DanielDent opened the pull request:
Adds support for the POSTGRES_XLOG_DIR environment variable, which specifies where the postgres transaction log is stored.

For some use cases, being able to place the transaction log on a different volume is useful.

(Existing support for providing flags via $POSTGRES_INITDB_ARGS is inadequate because of the need to create and chown/chmod the directory prior to running initdb)

DanielDent commented:
I just rebased on your latest changes, tests still pass. Please let me know if there is anything I can do to help get this merged.

tianon commented:
When it's used, isn't this directory something the user ought to be persisting as well?

DanielDent commented:
Definitely. I didn't add another VOLUME because VOLUMEs can't be undefined and not everyone will want multiple VOLUMEs for their data.

Additionally, users don't necessarily need another VOLUME to use this.

A user could define PGDATA to /var/lib/postgresql/data/pg_data, and POSTGRES_XLOG_DIR to /var/lib/postgresql/data/pg_xlog, and both would be persisted within that volume.

But it also gives them the option to attach different volumes to these mount points, or to volumes specified at runtime. The transaction log and the data files have different performance requirements/optimizations, and it's useful to be able to put them on mountpoints which take advantage of the required characteristics.

tianon commented:
So in the case of the xlogdir being a subdirectory of `$PGDATA`, permissions are already covered by the existing code, right?

I'm kind of inclined to put the permissions responsibility for this in the hands of the user, since it's already possible to do so today without explicit support for this functionality.  To put it another way, if we accept this knob for this directory/flag, which `initdb` or `postgres` flag is going to come next?

DanielDent commented:
The reason I created this fork/pull request is that I didn't find a way to have this work without modifying the upstream image. 

I initially tried having xlogdir as a subdirectory of $PGDATA. Postgres refuses to allow initdb to occur unless the data directory is empty. This creates problems if you are trying to mount a different volume onto that directory, because then the xlogdir directory exists and it's not an empty directory.

I certainly understand the general concern about not wanting to cover every permutation of every use case. That's where things like POSTGRES_INITDB_ARGS allow users to customize the image for themselves at runtime. 

Once users start having to manually fiddle with permissions just to launch a container, docker ceases to be something which removes administrative burden in the way it ought to, and I don't see away around it in this case.

DanielDent commented:
@tianon I just rebased again. 

I understand the desire to not have complicated support for every potential permutation of how Postgres is used, but I'd also point out that creating directories/setting permissions automatically so postgres can start without intervention does seem to be something that is seen as worth doing already - it's just that the current code doesn't work if you are using a custom transaction log directory.

If my commit can be merged as-is, yay! Or if you have alternative suggestions for how I could support this use case, or a more generalized merge-worthy approach I could implement which would cover my use case, I would be pleased to work towards having something which is merge-worthy.

tianon commented on docker-entrypoint.sh:
```diff
@@ -36,6 +36,13 @@ if [ "$1" = 'postgres' ] && [ "$(id -u)" = '0' ]; then
+        chmod 700 "$POSTGRES_XLOG_DIR"
+        chown -R postgres "$POSTGRES_XLOG_DIR"
+        POSTGRES_INITDB_ARGS="$POSTGRES_INITDB_ARGS --xlogdir $POSTGRES_XLOG_DIR"
+    fi
```
Thanks for your patience and good-nature, @DanielDent :disappointed: :heart:

I'm not _100%_ sold, but I concede that we probably need something for this.  Here are my thoughts on the implementation:

- I'd like a blank space above the comment (so that the `/var/run/postgresql` lines are separated from these)
- I'd rather name the variable `POSTGRES_INITDB_XLOGDIR` so it makes it more clear what it interacts with (and maps directly to the option it enables)
- the `POSTGRES_INITDB_ARGS="$POSTGRES_INITDB_ARGS ..."` line should start with `export`,
  ie `export POSTGRES_INITDB_ARGS="$POSTGRES_INITDB_ARGS ..."`
- the four-space indent you've got should be tabs (matching the rest of the file) :+1:

Happy to take over for any of this if you'd prefer, but want to give you first dibs (and a chance to converse if what I've proposed doesn't make sense or isn't accurate :smile:) :heart:

DanielDent commented:
Great. I've incorporated your proposed chances.

Additionally, I reworked things slightly to support secrets (file_env), which did not exist when I originally made these changes. If that doesn't make sense to you I left things without those changes on the master-nofileenv branch.

tianon commented on docker-entrypoint.sh:
```diff
@@ -47,7 +57,6 @@ if [ "$1" = 'postgres' ]; then
 
 	# look specifically for PG_VERSION, as it is expected in the DB dir
 	if [ ! -s "$PGDATA/PG_VERSION" ]; then
-		file_env 'POSTGRES_INITDB_ARGS'
```
This one needs to stay here in case we didn't hit the `[ "$(id -u)" = '0' ]` block for the above `if` (it should be harmless to duplicate it -- the function takes care of making sure multi-invocation is safe).

tianon commented on docker-entrypoint.sh:
```diff
@@ -37,6 +37,16 @@ if [ "$1" = 'postgres' ] && [ "$(id -u)" = '0' ]; then
 	chown -R postgres /var/run/postgresql
 	chmod g+s /var/run/postgresql
 
+    # Create the transaction log directory before initdb is run (below) so the directory is owned by the correct user
```
This line appears to still be space-indented. :smile:

DanielDent commented:
Thanks. Re-pushed.

DanielDent commented:
@tianon ping

tianon commented:
Sorry for the delay @DanielDent; I thought of another edge case (users who use `--user` rather than our auto-step-down code), so I updated the PR to include that. :+1:

tianon commented:
Thanks for your patience!
